subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -102,61 +103,83 @@ def check_payments(self, payments: Iterable[SubscriptionPayment]):
+            return None
 
         # Create subscription payment. Subscription is created automatically.
         subscription_payment = SubscriptionPayment.objects.create(
             provider_codename=self.codename,
-            provider_transaction_id=single_in_app.transaction_id,
+            provider_transaction_id=receipt_info.transaction_id,
+            # NOTE(kkalinowski): from my understanding, we can only receive receipt after the person has paid
+            # but the transaction (from the perspective of the app) might have not yet finished. Money were spent.
             status=SubscriptionPayment.Status.COMPLETED,
             # In-app purchase doesn't report the money.
             # We mark it as None to indicate we don't know how much did it cost.
             amount=None,
-            user=request.user,
+            user=user,
             plan=plan,
-            subscription_start=single_in_app.purchase_date,
-            subscription_end=single_in_app.expires_date,
+            subscription_start=receipt_info.purchase_date,
+            # If the cancellation date is set, it means that the user was refunded and for whatever reason
+            # we weren't notified about this purchase.
+            subscription_end=receipt_info.cancellation_date or receipt_info.expires_date,
         )
         subscription_payment.subscription.auto_prolong = False
         subscription_payment.save()
 
-        # Return the payment.
-        return Response(
-            SubscriptionPaymentSerializer(subscription_payment).data,
-            status=HTTP_200_OK,
-        )
+        return subscription_payment
+
+    @transaction.atomic
+    def _handle_receipt(self, request: Request, payload: AppleReceiptRequest) -> Response:
+        receipt = payload.transaction_receipt
+
+        # Validate the receipt. Fetch the status and product.
+        receipt_data = self.api.fetch_receipt_data(receipt)
+        self._is_receipt_valid(receipt_data)
+
+        if not receipt_data.latest_receipt_info:  # Either None or empty list.
+            raise InvalidAppleReceiptError('No latest receipt info provided, no recurring subscriptions to check.')
+
+        latest_payment = None
+        for receipt_info in receipt_data.latest_receipt_info:
+            # We receive all the elements and check whether we've actually activated them.
+            payment = self._handle_single_receipt_info(request.user, receipt_info)
+            if payment is None:
+                continue
+
+            if latest_payment is None or payment.subscription_end > latest_payment.subscription_end:
+                latest_payment = payment
+
+        # Return the latest payment or empty object if the plans are not properly assigned.
+        data = {}
+        if latest_payment is not None:
+            data = SubscriptionPaymentSerializer(latest_payment).data
+        return Response(data, status=HTTP_200_OK)
```

Comment:
The convention is to return serialized payment data to the app. You cannot return `{}` with status 200.

Reply:
Returning 400 in case we get nothing to present.